thread_support/thread_support.c

```diff
@@ -0,0 +1,351 @@
+    bsplit_basecase_func_t basecase;
+    bsplit_merge_func_t merge;
+    size_t sizeof_res;
+    bsplit_init_func_t init;
+    bsplit_clear_func_t clear;
+    void * args;
+    slong a;
+    slong b;
+    slong basecase_cutoff;
+    slong thread_limit;
+    int flags;
+}
+flint_parallel_binary_splitting_t;
+
+static void
+_bsplit_worker(void * _args)
+{
+    flint_parallel_binary_splitting_t * args = (flint_parallel_binary_splitting_t *) _args;
+
+    flint_parallel_binary_splitting(args->res, args->basecase, args->merge, args->sizeof_res, args->init, args->clear, args->args, args->a, args->b, args->basecase_cutoff, args->thread_limit, args->flags);
+}
+
+void
+flint_parallel_binary_splitting(void * res, bsplit_basecase_func_t basecase, bsplit_merge_func_t merge,
+    size_t sizeof_res, bsplit_init_func_t init, bsplit_clear_func_t clear, void * args, slong a, slong b, slong basecase_cutoff, int thread_limit, int flags)
+{
+    basecase_cutoff = FLINT_MAX(basecase_cutoff, 1);
+
+    if (b - a <= basecase_cutoff)
+    {
+        basecase(res, a, b, args);
+    }
+    else
+    {
+        void * left, * right;
+        slong m = a + (b - a) / 2;
+        slong nw;
+        slong nw_save;
+        slong nt;
+        thread_pool_handle * threads;
+        TMP_INIT;
+
+        TMP_START;
+
+        if (flags & FLINT_PARALLEL_BSPLIT_LEFT_INPLACE)
+        {
+            left = res;
+            right = TMP_ALLOC(sizeof_res);
+
+            init(right, args);
+        }
+        else
+        {
+            left = TMP_ALLOC(2 * sizeof_res);
+            right = (void *) (((char *) left) + sizeof_res);
+
+            init(left, args);
+            init(right, args);
+        }
+
+        if (thread_limit <= 0)
+            thread_limit = flint_get_num_threads();
+
+        nt = thread_limit;
+        nw = flint_request_threads(&threads, 2); /* request one extra worker */
+
+        if (nw == 0)
+        {
+            flint_parallel_binary_splitting(left, basecase, merge, sizeof_res, init, clear, args, a, m, basecase_cutoff, thread_limit, flags);
+            flint_parallel_binary_splitting(right, basecase, merge, sizeof_res, init, clear, args, m, b, basecase_cutoff, thread_limit, flags);
+        }
+        else
+        {
+            flint_parallel_binary_splitting_t right_args;
+
+            nw_save = flint_set_num_workers(nt - nt / 2 - 1);
+
+            right_args.res = right;
+            right_args.basecase = basecase;
+            right_args.merge = merge;
+            right_args.sizeof_res = sizeof_res;
+            right_args.init = init;
+            right_args.clear = clear;
+            right_args.args = args;
+            right_args.a = m;
+            right_args.b = b;
+            right_args.basecase_cutoff = basecase_cutoff;
+            right_args.thread_limit = thread_limit;
+            right_args.flags = flags;
+
+            thread_pool_wake(global_thread_pool, threads[0], nt / 2 - 1, _bsplit_worker, &right_args);
```

Comment:
So if the user passes in `thread_limit=1`, we are still requesting one extra worker? And if we get it, waking it up with `max_workers = -1`?

Reply:
For clarification, passing `max_workers=n` here causes the worker threads to have their thread global variable `_flint_num_workers` set to n before they start doing any work, which causes `flint_get_num_threads()` to return `n+1`. So, it looks like `n=-1` could be a problem.
Relevant lines:
https://github.com/wbhart/flint2/blob/trunk/thread_pool/wake.c#L35
https://github.com/wbhart/flint2/blob/trunk/thread_pool/init.c#L29